### PR TITLE
Editor: Add short cut for camera operation and gizmo mode selection

### PR DESCRIPTION
**Pull request by carlcc**
Add some basic shortcut that can be found in most Editors

### CODE REVIEW

File: Source/Tools/Editor/Tabs/Scene/SceneTab.cpp [hunk 1]
```diff
@@ -848,7 +848,13 @@ void SceneTab::OnUpdate(VariantMap& args)
         if (component != nullptr)
         {
             if (mouseHoversViewport_)
+            {
+                if (isMode3D_) {
```
Comment (rokups) [inline Source/Tools/Editor/Tabs/Scene/SceneTab.cpp]: Style. Brace to the next line.

File: Source/Tools/Toolbox/Scene/DebugCameraController.h [hunk 2]
```diff
@@ -46,11 +46,14 @@ class URHO3D_TOOLBOX_API DebugCameraController : public LogicComponent
...
     void Update(float timeStep) override;
 
+    void SetSelection(const ea::vector<WeakPtr<Node>>* selection);
+
 protected:
     /// Current camera speed.
     float speed_ = 2.f;
     /// Current mouse sensitivity.
     float mouseSensitivity_ = 0.1f;
+    const ea::vector<WeakPtr<Node>>* sceneSelection_ = nullptr;
```
Comment (rokups) [inline Source/Tools/Toolbox/Scene/DebugCameraController.h]: both `sceneSelection_` and `SetSelection` should have comments as part of our coding style. If you really can't think of a comment or think it is pointless - add empty comment like so:
```cpp
///
const ea::vector<WeakPtr<Node>>* sceneSelection_ = nullptr;
```

**Comment by rokups** [comment]
Hey thanks for the PR! :+1: 

Could you please describe your changes? I would love to know what i should test for.

**Comment by carlcc** [comment]
1. The debug camera will response our WASD and rotating command only when mouse is invisible (i.e. Right mouse button down)
2. Add `E` and `Q` (now WASDEQ) to tell the debug camera to move up or down (under orld coordinate)
3. Add support for rotating camera around something: When something (nodes) is/are selected, we can rotate debug camera by move mouse while `right mouse button` and `left shift` down.

**Comment by rokups** [comment]
It works wonderfully :heart: However these changes lost one feature. There was a very basic speeding camera up using shift. Now shift does something else. It would be great if you could also re-add speeding camera up. It could be a control next to buttons on scene viewport + adjusting speed using mouse wheel.

Also you forgot a comment above `void SetSelection()` ;)

A bonus task: at the moment using shift to rotate around object snaps to a new position immediately. It would be so awesome if camera smoothly moved into a new position. This is totally optional and if you do not feel like it - please ignore this suggestion ;)

**Comment by carlcc** [comment]
I tried, it seems that `shift` still has the ability to speed camera up.

1. If no node is selected, `shift` remains speed camera up.

2. If something is selected, `shift` performs rotate camera. 

If you don't like this, I think we can use another modify key such as `left alt`. In fact, my first thought is to use `left alt`, and many editors I have ever seen use it, but it seems that you had use it to toggle `traverse UI`. So... do you have a better modify key?

Uh... I found a bug: then holding shift, the camera direction is unchangable, I will fix it later.

As for `make  camera smoothly moved into a new position`. Indeed it provides much better user experience,  but I think it is not necessary, add this feature would make the code a little messy, we'd better enhance this feature in our future work. Maybe I will open another PR to implement this.

**Comment by rokups** [comment]
We should definitely no longer use shift for speedup as it produces very different behaviors in almost identical contexts. It even looks like a bug when node is selected. User presses shift and location rapidly changes while user expected camera to speed up.

What applications use alt? I have never encountered such. Shift also usually means speed boost in games so that felt natural.

We usually try to follow the rule: do what Unity does. In this case user can modify camera movement speed during camera movement only, by using mouse wheel. On-screen notification is shown about current speed, but i guess it would be enough for us to only have the behavior for now.

**Comment by nem0** [comment]
In unity shift == speedup, it's also true for most engines I worked with.

**Comment by carlcc** [comment]
Yes, both you are right. We'd wetter do not use shift to rotate camera. Almost all editors I have seen use it for speedup camera, too.


`What applications use alt`: 
UE4Editor:  LMB down + alt down + mouse move 
blender(with simulate 3rd mouse button on): MBB down + alt + mouse move
Xenko: same as UE4Editor
Maya/3dmax: I have not used them, but they might use the same logic as UE4Editor, according to UE4's docs.

Other editors such as CryEngine Sandox are not available for me at the moment, I am not quite sure if they use alt for rotating, but... perhaps.

**Comment by carlcc** [comment]
Changed the modify key to LALT and RALT, and change the WASDEQ's response condition from `Mouse Invisible` to `RMB down`.

I think this time, the exprience is much better, but I don't know how to cancel UI traversing :-( . I didn't find where the corresponding code is.

**Comment by rokups** [comment]
What UI traversing? Not sure what you have in mind here.

PR looks good. I think there is one final detail we need to address. `SetSelection()` takes externally owned pointer where we have no guarantees of it's lifetime. Although it probably does not happen in this case, it is possible that this pointer could be yanked from under `DebugCameraController` and cause a crash. Also center of selection code is duplicated in `Gizmo` class.

I am thinking we could extract center calculation code into `Gizmo::GetCenter(const ea::vector<WeakPtr<Node>>& nodes)`/`Gizmo::GetSelectionCenter()` and have `DebugCameraController::SetRotationCenter(const Vector3&)`/`DebugCameraController::ClearRotationCenter()`. Then there would be no way to misuse API by accident and cause crashes.

Let me know if you will have any other features, otherwise i will merge this PR after last thingy is cleared.

**Comment by carlcc** [comment]
image

As shown above, after releasing ALT, the 'Scene' tab was outlined

---

Agreed, CameraController cares about where to rotate around instead of what are selected, we only expose the center point to the controller is reasonable. Replace SetSelection withSetRotationCenter instead.

**Comment by rokups** [comment]
Oh yes that outline is ImGui keyboard navigation thingy. It is activated by pressing alt. I do not see a way to disable it for the next frame. I will bring it up to Omar and see if he knows of a possible solution. We can ignore it for now i think.

**Comment by rokups** [comment]
I talked to Omar regarding alt issue. Current temporary workaround is to add ` g.NavWindowingToggleLayer = true;` when alt is handled by our application. You could try that maybe. That aside - i think we are pretty much done with the PR. I can merge it after nav thingy, right?

**Comment by carlcc** [comment]
Yep. Now, the SystemUI will not be outlined if we perform an orbital rotation around something. Just merge. XP

**Comment by rokups** [comment]
Thank you very much :)

**Pull request merged**